### ISSUE TITLE
PositionalEncoding changes to dict during the runtime

### ISSUE THREAD

**Issue by aligholami** [allenai/allennlp#3493]
Hi. I got a problem with performing a simple `PositionalEncoding` on top of my embeddings. 
Here is my code for `model.py`:
```
from typing import Dict

import torch

from allennlp.models import Model
from allennlp.data import Vocabulary
from allennlp.modules import TextFieldEmbedder
from allennlp.modules.seq2seq_encoders.bidirectional_language_model_transformer import PositionalEncoding
from allennlp.modules.seq2seq_encoders import MultiHeadSelfAttention
from allennlp.modules import Seq2VecEncoder
from allennlp.modules import FeedForward

from allennlp.nn import InitializerApplicator
from allennlp.nn import util

from tcc.evaluation.avg_auc import AverageAUC


@Model.register('m2')
class MLPClassifier(Model):
    def __init__(self,
                 vocab: Vocabulary,
                 comment_embedder: TextFieldEmbedder,
                 positional_encoding: PositionalEncoding,
                 multi_head: MultiHeadSelfAttention,
                 seq_to_vec_encoder: Seq2VecEncoder,
                 fc: FeedForward,
                 initializer: InitializerApplicator = InitializerApplicator()) -> None:
        super().__init__(vocab)

        self.comment_embedder = comment_embedder
        self.positional_encoding = positional_encoding
        self.multi_head = multi_head
        self.seq_to_vec_encoder = seq_to_vec_encoder
        self.fc = fc

        self.num_classes = self.vocab.get_vocab_size("labels")
        self.metrics = {
            "accuracy": AverageAUC()
        }
        self.loss = torch.nn.BCEWithLogitsLoss()
        initializer(self)

    def forward(self,
                comment: Dict[str, torch.LongTensor],
                label: torch.LongTensor = None
                ) -> Dict[str, torch.Tensor]:

        comment_embedded = self.comment_embedder(comment)
        pos_encoded = self.positional_encoding(comment_embedded)
        multi_headed = self.multi_head(pos_encoded)
        comment_mask = util.get_text_field_mask(multi_headed)
        comment_encoded = self.seq_to_vec_encoder(multi_headed, comment_mask)

        logits = self.fc(comment_encoded)
        results_dict = {"logits": logits}

        if label is not None:
            loss = self.loss(logits, label)
            results_dict["loss"] = loss

            acc_metric = self.metrics["accuracy"]
            acc_metric(logits, label)

        return results_dict

    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
        return {metric_name: metric.get_metric(reset) for metric_name, metric in self.metrics.items()}
```

and here is my `config.json` file:

```
{
  "dataset_reader": {
    "type": "wikipedia_comments"
  },
  "train_data_path": "train",
  "validation_data_path": "val",
  "model": {
    "type": "m2",
    "comment_embedder": {
      "token_embedders": {
        "tokens": {
          "type": "embedding",
          "pretrained_file": "https://s3-us-west-2.amazonaws.com/allennlp/datasets/glove/glove.6B.100d.txt.gz",
          "embedding_dim": 100,
          "trainable": false
        }
      }
    },
    "positional_encoding": {
      "input_dim": 100
    },
    "multi_head":{
      "num_heads": 10,
      "input_dim": 100,
      "attention_dim": 300,
      "values_dim": 100
    },
    "seq_to_vec_encoder": {
      "type": "lstm",
      "bidirectional": true,
      "input_size": 100,
      "hidden_size": 100,
      "num_layers": 5,
      "dropout": 0.2
    },
    "fc": {
      "input_dim": 244,
      "num_layers": 3,
      "hidden_dims": [300, 100, 7],
      "activations": ["relu", "relu", "linear"],
      "dropout": [0.2, 0, 0]
    }
  },
  "iterator": {
    "type": "bucket",
    "sorting_keys": [["comment", "num_tokens"]],
    "batch_size": 256
  },
  "trainer": {
    "num_epochs": 40,
    "patience": 10,
    "cuda_device": 0,
    "grad_clipping": 5.0,
    "validation_metric": "+accuracy",
    "optimizer": {
      "type": "adagrad"
    }
  }
}

```
and I get the following error, which is very wired:

```
Traceback (most recent call last):
  File "/local-scratch/anaconda3/envs/TCC/bin/allennlp", line 8, in <module>
    sys.exit(run())
  File "/local-scratch/anaconda3/envs/TCC/lib/python3.7/site-packages/allennlp/run.py", line 18, in run
    main(prog="allennlp")
  File "/local-scratch/anaconda3/envs/TCC/lib/python3.7/site-packages/allennlp/commands/__init__.py", line 102, in main
    args.func(args)
  File "/local-scratch/anaconda3/envs/TCC/lib/python3.7/site-packages/allennlp/commands/train.py", line 124, in train_model_from_args
    args.cache_prefix)
  File "/local-scratch/anaconda3/envs/TCC/lib/python3.7/site-packages/allennlp/commands/train.py", line 168, in train_model_from_file
    cache_directory, cache_prefix)
  File "/local-scratch/anaconda3/envs/TCC/lib/python3.7/site-packages/allennlp/commands/train.py", line 252, in train_model
    metrics = trainer.train()
  File "/local-scratch/anaconda3/envs/TCC/lib/python3.7/site-packages/allennlp/training/trainer.py", line 478, in train
    train_metrics = self._train_epoch(epoch)
  File "/local-scratch/anaconda3/envs/TCC/lib/python3.7/site-packages/allennlp/training/trainer.py", line 320, in _train_epoch
    loss = self.batch_loss(batch_group, for_training=True)
  File "/local-scratch/anaconda3/envs/TCC/lib/python3.7/site-packages/allennlp/training/trainer.py", line 261, in batch_loss
    output_dict = self.model(**batch)
  File "/local-scratch/anaconda3/envs/TCC/lib/python3.7/site-packages/torch/nn/modules/module.py", line 541, in __call__
    result = self.forward(*input, **kwargs)
  File "./tcc/model/m2.py", line 50, in forward
    pos_encoded = self.positional_encoding(comment_embedded)
TypeError: 'dict' object is not callable
```

**Comment by brendan-ai2** [comment]
Hi @aligholami, it looks like `PositionalEncoding` doesn't  inherit from `Registrable` hence the underlying config is just being passed as a dictionary. This would be pretty easy to add to this file: https://github.com/allenai/allennlp/blob/3dda5ac9fcf390d8b83eb855249360711707c11c/allennlp/modules/seq2seq_encoders/bidirectional_language_model_transformer.py#L48 Happy to take a PR for this.

**Comment by ruleGreen** [comment]
Hello, Is this solved? I encountered this problem too. How to add Registrable at that file?

**Comment by h3lio5** [comment]
@brendan-ai2 I would like to take on this issue. Would the issue be solved by just inheriting the `Registrable` class?

**Comment by brendan-ai2** [comment]
@h3lio5 , yep that should pretty much be it. This is straightforward enough that we normally wouldn't add an automated test, but given that we forgot in the past, maybe that's a good idea here to prove that we've actually fixed the problem. Thanks for offering!

**Comment by h3lio5** [comment]
@brendan-ai2 I have made a PR. I hope I have done it right.

**Comment by DeNeutoy** [comment]
fixed by #3737
